src/structures/guild_preference.ts

```diff
@@ -89,35 +89,74 @@ export default class GuildPreference {
         this.gameOptions = options || { ...GuildPreference.DEFAULT_OPTIONS };
     }
 
-    static fromGuild(guildID: string, json?: GuildPreference): GuildPreference {
-        if (!json) {
-            return new GuildPreference(guildID, { ...GuildPreference.DEFAULT_OPTIONS });
-        }
-        // eslint-disable-next-line prefer-destructuring
-        const gameOptions = { ...json.gameOptions };
+    static validateGameOptions(gameOptions: GameOptions): GameOptions {
+        gameOptions = { ...gameOptions };
+
         // apply default game option for empty
-        let gameOptionModified = false;
         for (const defaultOption in GuildPreference.DEFAULT_OPTIONS) {
             if (!(defaultOption in gameOptions)) {
                 gameOptions[defaultOption] = GuildPreference.DEFAULT_OPTIONS[defaultOption];
-                gameOptionModified = true;
             }
         }
 
         // extraneous keys
         for (const option in gameOptions) {
             if (!(option in GuildPreference.DEFAULT_OPTIONS)) {
                 delete gameOptions[option];
-                gameOptionModified = true;
             }
         }
-        const guildPreference = new GuildPreference(guildID, gameOptions);
-        if (gameOptionModified) {
-            guildPreference.updateGuildPreferences();
+
+        return gameOptions;
+    }
+    static fromGuild(guildID: string, json?: GuildPreference): GuildPreference {
+        if (!json) {
+            return new GuildPreference(guildID, { ...GuildPreference.DEFAULT_OPTIONS });
         }
+        const gameOptions = this.validateGameOptions(json.gameOptions);
+
+        const guildPreference = new GuildPreference(guildID, gameOptions);
+        guildPreference.updateGuildPreferences();
+
         return guildPreference;
     }
 
+    async listPresets(): Promise<string[]> {
+        const presets = (await dbContext.kmq("game_option_presets")
+            .select(["preset_name"])
+            .where("guild_id", "=", this.guildID))
+            .map((x) => x["preset_name"]);
+        return presets;
+    }
+    async deletePreset(presetName: string): Promise<boolean> {
+        const result = await dbContext.kmq("game_option_presets")
+            .where("guild_id", "=", this.guildID)
+            .andWhere("preset_name", "=", presetName)
+            .del();
+        return result !== 0;
+    }
+
+    async savePreset(presetName: string) {
+        await dbContext.kmq("game_option_presets")
+            .insert({
```

Comment:
Just to confirm will this throw an error on duplicate preset insertion?

Reply:
Yes, due to the table.primary(["guild_id", "preset_name"]);